src/main/scala/mesosphere/util/ZookeeperDirectState.scala

```diff
@@ -0,0 +1,330 @@
+    handleExpunge(variable.asLocal).asScalaFutureWrapper
+  }
+
+  /**
+    * Returns the children of our ZK node.
+    * Returns an instance of ScalaFutureSupport type containing a Scala-Future
+    */
+  override def names: JFuture[util.Iterator[String]] = {
+    handleNames.asScalaFutureWrapper
+  }
+
+  /**
+    * Fetches the data from ZK. Returns an instance of ScalaFutureSupport type containing a Scala-Future.
+    * Uses a given callback to convert Curator's result to a Promise and then a Scala Future.
+    */
+  def handleFetch(name: String, callback: GetDataCallbackPromise): SFuture[Variable] = {
+
+    if (name == null) {
+      throw new RuntimeException("Whoah, name is null!")
+    }
+
+    exists(name) match {
+      case java.lang.Boolean.TRUE  => client.getData.inBackground(callback).forPath(fullPath(name))
+      case java.lang.Boolean.FALSE => callback.success(new LocalVariable(name))
+    }
+
+    callback.future
+  }
+
+  /**
+    * Marshall a variable to an Entry and store it in ZK.
+    *
+    * First fetches a variable and then flat maps its Scala future
+    * as the second argument to store(LocalVariable, LocalVariable)
+    * so it can ensure the right version is being updated.
+    *
+    * Returns an instance of ScalaFutureSupport type containing a Scala-Future
+    */
+  private def handleStore(lv: LocalVariable): SFuture[Variable] = {
+
+    implicit class JFutureToSFuture[A](f: JFuture[A]) {
+      def asScala: SFuture[A] = f match {
+        case sf: ScalaFutureSupport[A] => sf.future
+      }
+    }
+
+    // Grab the latest persisted variable and flatMap its future to store()'s future
+    val persistedVariableFuture: SFuture[Variable] = fetch(lv.name).asScala
+    val storeFuture: SFuture[Variable] = persistedVariableFuture.flatMap { pv: Variable =>
+      Option(handleStore(lv, pv.asLocal)) getOrElse SFuture(null)
+    }
+
+    storeFuture
+  }
+
+  /**
+    * Marshall a variable to an Entry and store it in ZK.
+    * Compares the variable with the latest persisted one to prevent an obsolete
+    * variable from being persisted.
+    *
+    * Returns null if the new veriable cannot be persisted
+    */
+  private def handleStore(lv: LocalVariable, persistedVariable: LocalVariable): SFuture[Variable] = {
+
+    val inputMatchesPersistedVersion: Boolean = persistedVariable.storedUuid == lv.storedUuid
+
+    if (inputMatchesPersistedVersion) {
+
+      val entryData = StateProtos.Entry.newBuilder()
+        .setName(lv.name)
+        .setUuid(ByteString.copyFromUtf8(UUID.randomUUID().toString))
+        .setValue(ByteString.copyFrom(lv.value()))
+        .build()
+        .toByteArray
+
+      mkDir(Some(lv.name))
+
+      val callback = new SetDataCallbackPromise(lv.name)
+      client.setData().inBackground(callback).forPath(fullPath(lv.name), entryData)
+      callback.future
+    }
+    else {
+      log.error(s"Attempt was made to persist ${lv.name} from obsolete version")
+      null
```

Comment:
The JNI implementation returns a Future containing null. `Future.successful(null)` should be the way to go.